src/ts/eventdispatcher.ts

```diff
@@ -196,28 +197,76 @@ class EventListenerWrapper<Sender, Args> {
  */
 class RateLimitedEventListenerWrapper<Sender, Args> extends EventListenerWrapper<Sender, Args> {
 
-  private rateMs: number;
-  private rateLimitingEventListener: EventListener<Sender, Args>;
+  // Timout in MS where the last seen event is delayed
+  private static NEXT_EVENT_TIMEOUT_MS = 50;
 
-  private lastFireTime: number;
+  private readonly rateMs: number;
+  private readonly rateLimitingEventListener: EventListener<Sender, Args>;
+
+  // save last seen event attributes
+  private lastSeenEvent: any;
+
+  private rateLimitTimout: Timeout;
+  private waitAndFireLastTimout: Timeout;
+
+  // Boolean to track if rateLimit timeout is currently active
+  private isRateLimiting: boolean = false;
 
   constructor(listener: EventListener<Sender, Args>, rateMs: number) {
     super(listener); // sets the event listener sink
 
     this.rateMs = rateMs;
-    this.lastFireTime = 0;
 
-    // Wrap the event listener with an event listener that does the rate-limiting
+    // delegating the received event
+    const fireEvent = (sender: Sender, args: Args) => {
+      this.fireSuper(sender, args);
+    };
+
+    // starting limiting the events to the given value
+    const startRateLimiting = () => {
+      if (this.lastSeenEvent) {
+        this.isRateLimiting = true;
+        this.rateLimitTimout.start();
+      }
+    };
+
+    // We need to wait a short time-windows to allow a current event to get delegated instead of the last saved one.
+    // Otherwise the lastSaveEvent will always be triggered after the rateLimiting.
+    this.waitAndFireLastTimout = new Timeout(RateLimitedEventListenerWrapper.NEXT_EVENT_TIMEOUT_MS, () => {
```

Comment:
I don't understand `NEXT_EVENT_TIMEOUT_MS`. Why isn't it simply `rateMs`? Yes there's the description above but I don't understand this either.

Reply:
Let me try that again:

_Let's assume we have a rate limit of 5 seconds._
1. The first event comes in and will be delegated immediately.
2. The rate limit timer with 5 seconds will be started.
    1. Each following event within the 5 seconds will be saved as last-seen event.

**Now we have 2 use cases:**
1. Incoming events stops during the rate limit timer. ->
    1. The rate limit timer expires after 5 seconds
    2. The last-seen event will be triggered

2. Incoming events never stops ->
    1. The rate limit timer expires after 5 seconds
    2. We need to check if a 'real' event is coming in (within the timeout `NEXT_EVENT_TIMEOUT_MS `)
Without this check, it could happen that a 'real' event will come at the same time we want to fire the last-seen event and that lead to one extra event within the rate limit timout.

        1. Yes, a 'real' event came within the interval (maybe at 5.00001) 
            1. cancel firing last seen event and delegate 'real' event
            2. restart rate limit
        2. No, no 'real' event came -> send last seen event

Reply:
IMO case 2 is not a problem and can be ignored. The event arriving a 5.00001 missed the 5 second interval and thus goes into the next one, which is pefectly fine. We can't make the interval longer just to make sure, because there can always be an event which is just an epsilon too late, even after `NEXT_EVENT_TIMEOUT_MS` - what happens then?